bt/tasks/utility/bt_evaluate_expression.cpp

```diff
@@ -0,0 +1,168 @@
+}
+
+void BTEvaluateExpression::set_node_param(Ref<BBNode> p_object) {
+	node_param = p_object;
+	emit_changed();
+	if (Engine::get_singleton()->is_editor_hint() && node_param.is_valid()) {
+		node_param->connect(LW_NAME(changed), Callable(this, LW_NAME(emit_changed)));
+	}
+}
+
+void BTEvaluateExpression::set_input_include_delta(bool p_input_include_delta) {
+	if (input_include_delta != p_input_include_delta) {
+		processed_input_values.resize(input_values.size() + int(p_input_include_delta));
+	}
+	input_include_delta = p_input_include_delta;
+	emit_changed();
+}
+
+void BTEvaluateExpression::set_input_vars(const PackedStringArray &p_input_vars) {
+	if (input_vars.size() != p_input_vars.size()) {
+	}
+	input_vars = p_input_vars;
+	emit_changed();
+}
+
+void BTEvaluateExpression::set_input_values(const TypedArray<BBVariant> &p_input_values) {
+	if (input_values.size() != p_input_values.size()) {
+		processed_input_values.resize(p_input_values.size() + int(input_include_delta));
+	}
+	input_values = p_input_values;
+	emit_changed();
+}
+
+void BTEvaluateExpression::set_result_var(const String &p_result_var) {
+	result_var = p_result_var;
+	emit_changed();
+}
+
+//**** Task Implementation
+
+PackedStringArray BTEvaluateExpression::get_configuration_warnings() {
+	PackedStringArray warnings = BTAction::get_configuration_warnings();
+	if (expression_string.is_empty()) {
+		warnings.append("Expression string is not set.");
+	}
+	if (node_param.is_null()) {
+		warnings.append("Node parameter is not set.");
+	} else if (node_param->get_value_source() == BBParam::SAVED_VALUE && node_param->get_saved_value() == Variant()) {
+		warnings.append("Path to node is not set.");
+	} else if (node_param->get_value_source() == BBParam::BLACKBOARD_VAR && node_param->get_variable() == String()) {
+		warnings.append("Node blackboard variable is not set.");
+	}
+	return warnings;
+}
+
+void BTEvaluateExpression::_setup() {
+	parse();
+	ERR_FAIL_COND_MSG(is_parsed != Error::OK, "BTEvaluateExpression: Failed to parse expression: " + expression.get_error_text());
+}
+
+Error BTEvaluateExpression::parse() {
+	PackedStringArray processed_input_vars;
+	processed_input_vars.resize(input_vars.size() + int(input_include_delta));
+	String *processed_input_vars_ptr = processed_input_vars.ptrw();
+	if (input_include_delta) {
+		processed_input_vars_ptr[0] = "delta";
+	}
+	for (int i = int(input_include_delta); i < input_vars.size(); ++i) {
+		processed_input_vars_ptr[i] = input_vars[i];
+	}
+
+	is_parsed = expression.parse(expression_string, processed_input_vars);
+	return is_parsed;
+}
+
+String BTEvaluateExpression::_generate_name() {
+	String input_vars_str = input_include_delta ? "delta" : "";
+	if (input_vars.size() > 0) {
+		if (!input_vars_str.is_empty()) {
+			input_vars_str += ", ";
+		}
+		input_vars_str += vformat("%s", input_vars).trim_prefix("[").trim_suffix("]").replace("\"", "");
+	}
+	return vformat("EvaluateExpression %s with vars [%s]  node: %s  %s",
```

Comment:
Not sure here what is our consensus here. Do you agree that vars in the generated "name" should be removed?

Reply:
Oh, you mean removing the list itself `[%s]`)? I thought you meant just the `vars` string 😅.

Reply:
I don't really mind removing it entirely either.

Reply:
Let me make the change.

Reply:
Yeah, those variables should be present in the expression anyway. Seems like it's not necessary to have them also enumerated explicitly. If we're worried, that some of them go unused, we can still do that in get_configuration_warning maybe.

Reply:
> If we're worried, that some of them go unused, we can still do that in get_configuration_warning maybe.

Implementing that in `get_configuration_warning` may be hard though -- `Expression` has it's own parser and we'd have to either have it emit signals every time a token with `TK_IDENTIFIER` is found, or move the internal parser to the public interface and add hooks to it. I don't think any of those changes would be accepted upstream. Although we may be able to use the GDScript parser -- I can look into that later.

For what is worth, I initially implemented listing the vars for debugging purposes. The problem is that in the expression alone, each visible variable has two value contexts -- the value taken from the `Object` property the expression gets run against and the value from `input_values`, the latter of which only works when the variable name is specified in `input_names`. For example, in the expression `print(delta)`, `delta` can either be `Object.delta` or `input_value[0]` if `input_include_delta` is `true`. The reason I don't think it's necessary is because they can just look at the inspector.

Reply:
Thanks for explanation. Unless it can be done with a few lines of code, I don't think it's worth the effort.
It's good as it is, IMO. Let's keep it like this, and it can be improved later if necessary.